packages/opentelemetry-core/src/platform/node/id.ts

```diff
@@ -14,23 +14,27 @@
  * limitations under the License.
  */
 
-import * as crypto from 'crypto';
-
 const SPAN_ID_BYTES = 8;
 const TRACE_ID_BYTES = 16;
 
 /**
  * Returns a random 16-byte trace ID formatted/encoded as a 32 lowercase hex
  * characters corresponding to 128 bits.
  */
-export function randomTraceId(): string {
-  return crypto.randomBytes(TRACE_ID_BYTES).toString('hex');
-}
+export const randomTraceId = getIdGenerator(TRACE_ID_BYTES);
 
 /**
  * Returns a random 8-byte span ID formatted/encoded as a 16 lowercase hex
  * characters corresponding to 64 bits.
  */
-export function randomSpanId(): string {
-  return crypto.randomBytes(SPAN_ID_BYTES).toString('hex');
+export const randomSpanId = getIdGenerator(SPAN_ID_BYTES);
+
+const SHARED_BUFFER = Buffer.allocUnsafe(TRACE_ID_BYTES);
+function getIdGenerator(bytes: number): () => string {
```

Comment:
@dyladan I know it was already like that, but shouldn't the id generator ensure the returned value  has at least one non-zero byte? The [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/c33919ea1265081967287a3c3160ba823797bbce/specification/trace/api.md#spancontext) requires that for both `traceId` and `spanId`.

Reply:
We discussed this at one time in the past. Technically it is _possible_ for it to randomly return all 0's, but it is so vanishingly unlikely that I don't think it's worth the extra boolean check every time an id is generated. I can be talked out of this position though.

/cc @open-telemetry/javascript-maintainers

Reply:
Fair enough @dyladan . One could accept that in the unlikely event that happens then it could be equivalent for the trace to not being sampled.

But, is that what happens in case either `traceId` or `spanId` is 0? Is the SpanContext gracefully discarded?

I'm asking to understand whether propagators should explicitly check those fields before calling `setExtractedSpanContext`. (Note: The B3 and HTTP propagators do check it).

Reply:
We could create one random byte starting at one instead 0. Just one bit of randomness would be lost.

Reply:
What do you think of this?

```typescript
const SHARED_BUFFER = Buffer.allocUnsafe(TRACE_ID_BYTES);
function getIdGenerator(bytes: number): () => string {
  return function generateId() {
    for (let i = 0; i < bytes / 4; i++) {
      SHARED_BUFFER.writeUInt32BE((Math.random() * 2 ** 32) >>> 0, i * 4);
    }
    // Ensure the last bit is never zero to guarantee a valid w3c id is generated
    SHARED_BUFFER[bytes - 1] = SHARED_BUFFER[bytes - 1] | 0x1;
    return SHARED_BUFFER.toString('hex', 0, bytes);
  };
}
```

Reply:
LGTM